lib/RefBackend/RefBackend.cpp

```diff
@@ -119,35 +104,36 @@ static LogicalResult mungeFunction(
   }
 
   SmallVector<Operation *> toErase;
-  bool isCompatible = false;
+  bool isCompatible = true;
   func.walk([&](ReturnOp op) {
-    auto returnType = op.getOperandTypes()[0];
-
+    auto types = op.getOperandTypes();
     b.setInsertionPoint(op);
     // Memref Types.
-    if (auto memrefReturnType = returnType.dyn_cast<MemRefType>()) {
-      auto elemType = memrefReturnType.getElementType();
-      auto unRankedType = UnrankedMemRefType::get(elemType, 0);
-      isCompatible =
-          isReturnOpCompatible(op, consumeFuncReturnFuncs, unRankedType);
-      if (!isCompatible)
-        return;
-
-      // Cast to unranked memref type before sending it as a function argument.
-      auto cast = b.create<memref::CastOp>(
-          op.getLoc(), op.getOperand(0),
-          getAbiTypeForMemRef(op.getOperandTypes()[0]));
-      replaceCallToFunction(b, op, consumeFuncReturnFuncs[unRankedType],
-                            cast.getResult(), toErase);
-      // Elemental types.
-    } else if (returnType.isa<IntegerType>() || returnType.isa<FloatType>()) {
-      isCompatible =
-          isReturnOpCompatible(op, consumeFuncReturnFuncs, returnType);
-      if (!isCompatible)
-        return;
-      replaceCallToFunction(b, op, consumeFuncReturnFuncs[returnType],
-                            op->getOperand(0), toErase);
+    SmallVector<Type> retTypes;
+    SmallVector<Value> retVals;
+    for (auto en : llvm::enumerate(types)) {
+      Type returnType = en.value();
+      Value value = op.getOperand(en.index());
+      if (auto memrefReturnType = returnType.dyn_cast<MemRefType>()) {
+        auto elemType = memrefReturnType.getElementType();
+        returnType = UnrankedMemRefType::get(elemType, 0);
+        // Cast to unranked memref type before sending it as a function
+        // argument.
+        value = b.create<memref::CastOp>(
+            op.getLoc(), value, getAbiTypeForMemRef(op.getOperandTypes()[0]));
+      }
+      retTypes.push_back(returnType);
+      retVals.push_back(value);
     }
+
+    auto it = consumeFuncReturnFuncs.end();
```

Comment:
I would recommend that you change this to create the call with a systematically derived name (implement a function `getConsumeReturnFunctionNameForReturnTypes(TypeRange types) -> std::string`. Then save a `SmallVector<std::pair<std::vector<Type>, std::string>>` of consume return functions that need to be created. Finally, at the end of the pass, create the consume return functions that are needed.

Reply:
Added the name deriving function as suggested. I ended up storing the consume return functions that need to be created into a map std::map<std::string, std::vector<Type>> to avoid creating a function twice. Also, added a string set to keep track of supported consume return functions for error reporting.